So many updates

its-miroma opened the pull request:
> [!TIP]
> This diff page may have better performance than the review: <https://github.com/FabricMC/fabric-docs/compare/main...its-miroma:fmc:framework>

- Update `.github/CODEOWNERS`
- Update `.github/workflows`:
  - `build.yml`:
    - run on `main` only, otherwise pull requests from `l10n/main` run build twice
    - add conditions to avoid "Run failed" mails
  - `markdownlint.yml`:
    - remove `branches`, as `paths` already excludes pull requests from `l10n/main`
    - remove `workflow_dispatch`
  - `publish.yml`:
    - run on `ubuntu-24.04`
    - add conditions to avoid "Run failed" mails
- `settings.gradle`: build `latest` only
  - `update.ts`: don't add new versions to `settings.gradle`
- Add an `h1` heading from `$frontmatter.title` automatically
  - ~~remove all `# ` headings~~
  - `.markdownlint-cli2.yaml`: enable `MD025`
- add new `markdownlint` rules:
  - `max-one-sentence-per-line` (not in here)
  - `incorrect-mod-id`
  - `no-horizontal-rules`
  - `no-html-line-breaks`
  - `no-trailing-punctuation-in-lists
- enable lazy image loading
- add a footer
- format ~~many~~ few files with prettier
- add more vscode extension recommendations
- hide files that shouldn't be edited in vscode
- fix #272
- ~~fix formatting mistakes in `translated`~~
- ~~lint _all_ markdown files~~
- ~~delete `contributing.md` under `versions`~~

its-miroma committed:
Framework updates

- Update `.github/CODEOWNERS`
- `build.yml`: run on `main` only
  - otherwise pull requests from `l10n/main` run build twice
- `markdownlint.yml`: remove `branches`
  - `paths` already excludes pull requests from `l10n/main`
- `settings.gradle`: build `latest` only
- `update.ts`: don't add new versions to `settings.gradle`

its-miroma committed:
Update `.github/CODEOWNERS`

its-miroma committed:
Add conditions to avoid "Run failed" mails

its-miroma commented on .github/workflows/build.yml:
add this job to avoid weird emails:

```yaml
      - name: Cleanup
        run: exit 0
```

its-miroma commented on .github/workflows/publish.yml:
add this job to avoid weird emails:

```yaml
      - name: Cleanup
        run: exit 0
```

its-miroma committed:
Add "Cleanup" job

its-miroma committed:
Fix `if` condition

its-miroma commented on .github/workflows/publish.yml:
```diff
@@ -32,6 +33,7 @@ jobs:
           path: "./.vitepress/dist"
 
   deploy:
```
This step is running on `ubuntu-22.04`, is this intentional?

modmuss50 replied:
No, this can be updated.

its-miroma committed:
Upgrade `runs-on` to `ubuntu-24.04`

its-miroma committed:
add a footer

its-miroma committed:
add new markdownlint rules

- `incorrect-mod-id`
- `no-horizontal-rules`
- `no-html-line-breaks`
- `no-trailing-punctuation-in-lists`
- TODO: `max-one-sentence-per-line`

---

- don't lint `versions` at all

its-miroma committed:
enable lazy loading of images

its-miroma committed:
add more recommended extensions

its-miroma committed:
add `h1` heading automatically from frontmatter

its-miroma committed:
prepare for localization in `update`

its-miroma commented on .markdownlint-cli2.yaml:
```diff
@@ -122,11 +139,17 @@ config:
         replace: ""
         searchScope: code
 customRules:
+  # TODO: add      markdownlint-rule-max-one-sentence-per-line/rule.js
```
this would be helpful for a few reasons:

- no longer worrying about line wraps
- longer/complicated sentences stick out
- crowdin doesn't show a confusing new line in the middle of source strings

it would however lead to a monstrous diff, so i avoided doing it here.

if i ever wanted to add it, the best way would be to render the markdown files after the sentences split and run a `diff` on the built files to check for broken/changed formatting

its-miroma commented on .vitepress/config.mts:
```diff
@@ -30,18 +30,21 @@ export default defineVersionedConfig(
         // Use the snippet plugin (transclusion, etc.)
         md.use(snippetPlugin);
       },
+      image: {
+        lazyLoading: true,
+      },
```
would appreciate feedback/testing on this

modmuss50 replied:
Id argue we dont need this? By the time you scroll down the image it will have already been loaded.

its-miroma replied:
Lazy loading of images allows faster loading of a page because images get loaded when needed.

Check out <https://developer.mozilla.org/en-US/docs/Web/Performance/Lazy_loading>

its-miroma commented on .vitepress/i18n.ts:
```diff
@@ -220,6 +220,14 @@ function generateTranslatedThemeConfig(localeCode: string): Fabric.ThemeConfig {
+    footer: {
+      copyright: websiteResolver("footer.copyright").replace(
+        "%s",
+        `<a href=\"https://github.com/FabricMC/fabric-docs/blob/main/LICENSE\" target=\"_blank\" rel=\"noreferrer\">${websiteResolver("footer.license")}</a>`
```
unsure what the best way to sanitize `footer.license` is

a malicious translation could close the `</a>`...

modmuss50 replied:
Maybe use `document.createElement("a")` with ``.textContent and then get the `outerHTML` Im not sure if this works in this context though?

its-miroma commented on .vitepress/theme/index.ts:
```diff
@@ -16,26 +16,33 @@ import "./style.css";
     const children = {
-      "doc-before": () => h(VersionReminder),
+      "doc-before": () => [
+        frontmatter.value.title
+          ? h("h1", { class: "vp-doc" }, frontmatter.value.title)
+          : null,
```
is there ever a need for two different titles in `<head>` and `<body>`?

IMB11 replied:
Yes, the homepage

its-miroma replied:
`doc-before` (line 35) only includes doc pages, not the homepage nor the 404 page

its-miroma commented on .vitepress/theme/style.css:
```diff
@@ -8,7 +8,7 @@
   unicode-range: U+1F1E6-1F1FF, U+1F3F4, U+E0062-E0063, U+E0065, U+E0067,
     U+E006C, U+E006E, U+E0073-E0074, U+E0077, U+E007F;
-  src: url("./CountryFlagsPolyfill.woff2") format('woff2');
+  src: url("./CountryFlagsPolyfill.woff2") format("woff2");
```
if we used Noto Color Emoji we could embed it without shipping the binary file in the repo

IMB11 replied:
We dont need to include all emojis though, we only need the flags.

its-miroma commented on contributing.md:
```diff
@@ -3,8 +3,6 @@ title: Contribution Guidelines
 ---
 
-# Contribution Guidelines {#contributing}
-
```
every single markdown file will have at least a two-line diff.

I've also deleted some `contributing` files which slipped through in `versions`

its-miroma reviewed (commented):
~~this PR may be unreviewable~~

its-miroma committed:
fix #272

its-miroma committed:
override default margin of VPHome

its-miroma committed:
npm audit fix

modmuss50 commented on .github/workflows/build.yml:
```diff
@@ -37,3 +43,8 @@ jobs:
+  cleanup:
+    runs-on: ubuntu-24.04
+    steps:
+      - run: exit 0
```
Why is this needed?

its-miroma replied:
See this message on Discord: <https://discord.com/channels/507304429255393322/1208846408552030238/1331740336640233584>

TL;DR: not running any job still sends annoying mail

modmuss50 commented on .github/CODEOWNERS:
```diff
@@ -1,8 +1,18 @@
 * @FabricMC/documentation
 
+.gitattributes @FabricMC/developers
```
I think we should changed this file to default to devs, but allow the docs team access to the content.

modmuss50 commented on .github/workflows/build.yml:
```diff
@@ -1,12 +1,17 @@
+
+on:
+  push:
+    branches: ["main"]
```
Why would we only want to build the main branch?

its-miroma replied:
Because `pull_request` (line 6) already builds the docs for incoming contributions

modmuss50 replied:
Again, this doesnt match what we have for other repos, so I dont get whats special here?

its-miroma replied:
If this is done on other repos as well, then the same applies there. Some builds are redundant, and these are not any more special than the others'. If you don't believe it's worth fixing we can revert these changes.

modmuss50 replied:
I get why we wouldnt want to try to deploy from other branches, but it should still attempt to build. Can you disable actions on your branch if you dont want to waste builds (at Githubs expense not yours).

its-miroma replied:
Why would we want to build the docs and the mod on branches if a `pull_request` is not open?
Also, one could say that Netlify builds the docs, so that entire job is redundant.

I get why you'd be reluctant to change these actions, just confirm that I should undo these if needed

> Can you disable actions on your branch if you dont want to waste builds

I'm not sure how I can do that...

> at Githubs expense not yours

at FabricMC's minutes expense, I don't know how much of a problem that is

modmuss50 commented on .github/workflows/build.yml:
```diff
@@ -1,12 +1,17 @@
 
 jobs:
   docs:
+    if: github.repository == 'FabricMC/fabric-docs'
```
This doesnt make any sense for building, if you fork the repo the actions should still run. Only the publish task needs this.

its-miroma replied:
See explanation following this message on Discord: <https://discord.com/channels/507304429255393322/1208846408552030238/1332670446818299944>

TL;DR: syncing a fork's main branch would trigger the job uselessly and cause it to fail and send annoying notifications.

modmuss50 replied:
I disagree with it being useless, it should succeed. We dont do this with any other projects so I dont understand why its an issue here.

its-miroma replied:
That's my point: it _will_ most certainly succeed because syncing a repo's `main` branch just means pulling the (successful) commits from upstream.

You can also consider it to be a security risk, because if it is ran on updates to `main` on forks, a malicious commit could edit the action, and the new action might get ran without any approval (I'd need to test it to confirm these suspicion).

modmuss50 replied:
> a malicious commit

The malicious commit would need to make it into our main branch, that should never happen. And could only happen if someone approved the PR.

its-miroma replied:
that's reassuring; that still leaves the first part relevant

modmuss50 replied:
I dont see a major problem with that, in this one edge case it will build something that is known to already build, seems like a very minor issue that isnt our problem.

modmuss50 commented on .github/workflows/publish.yml:
```diff
@@ -16,6 +16,7 @@ permissions:
 
 jobs:
   build:
+    if: github.repository == 'FabricMC/fabric-docs'
```
Only needed for the deploy job.

its-miroma replied:
See explanation above

modmuss50 commented on .markdownlint-cli2.yaml:
```diff
@@ -9,16 +11,15 @@ config:
   MD029:
     style: ordered
-  MD033: false
+  MD033: false # TODO: remove
```
TODO? Is this for the future?

its-miroma replied:
That's the no HTML tags rule, it's possible to set an allowlist which may be safer, but I don't know if that's worth it

its-miroma committed:
Update `.github/CODEOWNERS`

its-miroma committed:
some formatting

its-miroma committed:
revert changes to workflows

its-miroma committed:
configure `allowed_elements`

modmuss50 commented:
Looks good, we can either merge this and then get the lint fixes, or I can just slap a re approval with the lint fixes as part of this. Thanks for splitting it out. In the future please try to keep PRs to just 1 or two changes.

its-miroma commented:
let's merge this PR and keep it clean, unless @IMB11 wants to review it.
The linting fixes are in #282
